lib/Doctrine/DBAL/Connections/PrimaryReplicaConnection.php

```diff
@@ -0,0 +1,361 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function executeUpdate($query, array $params = [], array $types = []): int
+    {
+        $this->switchToPrimary();
+
+        return parent::executeUpdate($query, $params, $types);
+    }
+
+    public function beginTransaction(): bool
+    {
+        $this->switchToPrimary();
+
+        return parent::beginTransaction();
+    }
+
+    public function commit(): bool
+    {
+        $this->switchToPrimary();
+
+        return parent::commit();
+    }
+
+    public function rollBack(): bool
+    {
+        $this->switchToPrimary();
+
+        return parent::rollBack();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function delete($tableName, array $identifier, array $types = []): int
+    {
+        $this->switchToPrimary();
+
+        return parent::delete($tableName, $identifier, $types);
+    }
+
+    public function close(): void
+    {
+        unset($this->primary, $this->replica);
+
+        parent::close();
+
+        $this->_conn = null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function update($tableName, array $data, array $identifier, array $types = []): int
+    {
+        $this->switchToPrimary();
+
+        return parent::update($tableName, $data, $identifier, $types);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function insert($tableName, array $data, array $types = []): int
+    {
+        $this->switchToPrimary();
+
+        return parent::insert($tableName, $data, $types);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function exec($statement): int
+    {
+        $this->switchToPrimary();
+
+        return parent::exec($statement);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function createSavepoint($savepoint): void
+    {
+        $this->switchToPrimary();
+
+        parent::createSavepoint($savepoint);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function releaseSavepoint($savepoint): void
+    {
+        $this->switchToPrimary();
+
+        parent::releaseSavepoint($savepoint);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function rollbackSavepoint($savepoint): void
+    {
+        $this->switchToPrimary();
+
+        parent::rollbackSavepoint($savepoint);
+    }
+
+    public function query(): Statement
+    {
+        $this->switchToPrimary();
+        assert($this->_conn instanceof DriverConnection);
+
+        $args = func_get_args();
+
+        $logger = $this->getConfiguration()->getSQLLogger();
+        if ($logger) {
+            $logger->startQuery($args[0]);
+        }
+
+        $statement = $this->_conn->query(...$args);
+
+        $statement->setFetchMode($this->defaultFetchMode);
+
+        if ($logger) {
+            $logger->stopQuery();
+        }
+
+        return $statement;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function prepare($statement): Statement
+    {
+        $this->switchToPrimary();
```

Comment:
This primary/replica thing in DBAL is a really poor and leaky abstraction. Why do all statements need to be prepared on the primary endpoint?

Reply:
:thinking: would this be the reason: https://github.com/doctrine/dbal/blob/0ad4c165318128c50d9a425c8c15520c094f465b/lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php#L43-L47

Not sure what it refers to.

Reply:
We cannot detect if a query is truely only a read operation, it would require parsing the query string. 

ORM uses executeUpdate() or exec(), or directly prepare()/execute() to do writes. 

But executeQuery() is semantically about reads, so we can be sure that when this method is used, its a almost 100% read operation (not 100%, because users can just perform a write query with executeQuery even though it doesnt make sense).